build: Do not ignore Homebrew's SQLite on macOS

hebasto opened the pull request:
On master (7ae86b3c6845873ca96650fc69beb4ae5285c801) installed Homebrew `sqlite` package is ignored during build on macOS.

This PR fixes this issue and update macOS build docs.

Closes #20498.

hebasto commented:
Mind considering backporting into 0.21?

jonasschnelli commented:
utACK ef7cd61a401f0c51afa0cb8c46271e2d7e9e9424

jonasschnelli commented:
Needs rebase (due to #20478)

hebasto commented:
Rebased ef7cd61a401f0c51afa0cb8c46271e2d7e9e9424 -> 0479464ac938b3d51f7f38b19a20ed2a5dcf9c60 ([pr20527.01](https://github.com/hebasto/bitcoin/commits/pr20527.01) -> [pr20527.02](https://github.com/hebasto/bitcoin/commits/pr20527.02)) due to the conflict with #20478.

willcl-ark reviewed (changes requested):
Tested 0479464ac938b3d51f7f38b19a20ed2a5dcf9c60 on macOS 11.0.1 with Homebrew 2.6.0

Verified that homebrews `sqlite` is auto-selected after installation and builds sucessfully.

Interestingly, after uninstalling brews `sqlite` I still found references to ghost homebrew libraries in the Makefile because of the way `brew --prefix sqlite3` works: https://github.com/Homebrew/brew/blob/3821d37997a7b385cc97bb51169de56f641d5c39/Library/Homebrew/cmd/--prefix.rb#L23

It means that if `brew` is installed but brew's `sqlite` is not, then `brew --prefix sqlite` will still return where it _would_ install it, and having this path added to the Makefile accordingly.

I have tested that installation does indeed still pass in this case, but it might still be worth testing the sqlite prefix and then the presence of the binary/bin folder itself to avoid a future issue, e.g.:

`test ! -d $(brew --prefix sqlite)/bin`

laanwj commented on doc/build-osx.md:
````diff
@@ -19,7 +19,7 @@ Then install [Homebrew](https://brew.sh).
+brew install automake libtool boost miniupnpc pkg-config python qt libevent qrencode
 ```
 
 If you run into issues, check [Homebrew's troubleshooting page](https://docs.brew.sh/Troubleshooting).
````
It might be helpful to add a general note here that wallet support requires one or both of the dependencies in the sections below.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/20527#issuecomment-738075686).

hebasto commented:
@willcl-ark 
> It means that if `brew` is installed but brew's `sqlite` is not, then `brew --prefix sqlite` will still return where it _would_ install it, and having this path added to the Makefile accordingly.

I think the code in this PR handles this case fine, as it actually adds a non-existent path to the `PKG_CONFIG_PATH`, that in turn effectively is noop, right?

willcl-ark commented:
> @willcl-ark
> 
> > It means that if `brew` is installed but brew's `sqlite` is not, then `brew --prefix sqlite` will still return where it _would_ install it, and having this path added to the Makefile accordingly.
> 
> I think the code in this PR handles this case fine, as it actually adds a non-existent path to the `PKG_CONFIG_PATH`, that in turn effectively is noop, right?

I'm not that experienced with Makefiles' path interpreter but agree that having that extra (non-existant) path in `PKG_CONFIG_PATH` did not seem to affect the build in any way when brews `sqlite3` was not installed.

hebasto commented:
Updated 0479464ac938b3d51f7f38b19a20ed2a5dcf9c60 -> 7bf449ee5c4fb0748370981f1205c356162e58b4 ([pr20527.02](https://github.com/hebasto/bitcoin/commits/pr20527.02) -> [pr20527.03](https://github.com/hebasto/bitcoin/commits/pr20527.03), [diff](https://github.com/hebasto/bitcoin/compare/pr20527.02..pr20527.03)):

- addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/20527#discussion_r535202454):
> It might be helpful to add a general note here that wallet support requires one or both of the dependencies in the sections below.

jonasschnelli reviewed (approved):
code review ACK 7bf449ee5c4fb0748370981f1205c356162e58b4

hebasto committed:
build, refactor: Check that Homebrew's qt5 package is actually installed

This change unifies Homebrew packages workflow, and does not change
behavior.

hebasto committed:
build: Use Homebrew's sqlite package if it is available

hebasto committed:
doc: Update wallet database installation guide for macOS

hebasto commented:
Updated 7bf449ee5c4fb0748370981f1205c356162e58b4 -> 03c65e4f704c4baa9bd72a56e8afdb92ec5d4d33 ([pr20527.03](https://github.com/hebasto/bitcoin/commits/pr20527.03) -> [pr20527.04](https://github.com/hebasto/bitcoin/commits/pr20527.04)):
- rebased due to the conflict with #20563
- used uniform approach to check if Homebrew's package is actually installed, that effectively implements @willcl-ark's [suggestion](https://github.com/bitcoin/bitcoin/pull/20527#pullrequestreview-543914043)

willcl-ark commented:
tACK 03c65e4f704c4baa9bd72a56e8afdb92ec5d4d33

Building without brew `sqlite3` (but with brew installed) now leaves no trace in config.log:

```
SQLITE_CFLAGS=''
SQLITE_LIBS='-lsqlite3'
```

vs building with brew `sqlite3`:

```
SQLITE_CFLAGS='-I/usr/local/Cellar/sqlite/3.33.0/include'
SQLITE_LIBS='-L/usr/local/Cellar/sqlite/3.33.0/lib -lsqlite3'
```

laanwj commented on configure.ac:
```diff
@@ -642,16 +642,19 @@ case $host in
-           PKG_CONFIG_PATH="$qt5_prefix/lib/pkgconfig:$PKG_CONFIG_PATH"
-           export PKG_CONFIG_PATH
+
+         if $BREW list --versions sqlite3 >/dev/null; then
```
Shouldn't this, in principle, be conditional on sqlite3 support actually being requested?

I'm sure it works but I think it leaves some scope for unexpected differences in behavbior if the build configuration unconditionally depends on the system instead of what the user asks.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/20527#issuecomment-739860794).

hebasto commented:
Updated 03c65e4f704c4baa9bd72a56e8afdb92ec5d4d33 -> c932e0d67e4b369e4265267da6c8bebac2b6fb53 ([pr20527.04](https://github.com/hebasto/bitcoin/commits/pr20527.04) -> [pr20527.05](https://github.com/hebasto/bitcoin/commits/pr20527.05), [diff](https://github.com/hebasto/bitcoin/compare/pr20527.04..pr20527.05)):

- addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/20527#discussion_r537376483):
> Shouldn't this, in principle, be conditional on sqlite3 support actually being requested? (like the `use_bdb` above, and same for `qt5` below)
> 
> I'm sure it works but I think it leaves some scope for unexpected differences in behavbior if the build configuration unconditionally depends on the system instead of what the user asks.

- (minor) improved robustness of `use_bdb` check

willcl-ark commented:
With brew sqlite installed, running `./configure --with-sqlite=no` I still see a single artefact of sqlite in config.log:

```
5390:PKG_CONFIG_PATH='/usr/local/opt/qt/lib/pkgconfig:/usr/local/opt/sqlite/lib/pkgconfig:
```

It does not appear to impact the build, but I am curious how it's ending up there, as the code would suggest this would only be prepended to PKG_CONFIG_PATH when `if test "x$use_sqlite" != xno && $BREW list --versions sqlite3 >/dev/null` is satisfied?

That said, another tACK of c932e0d67e4b369e4265267da6c8bebac2b6fb53

hebasto commented:
> With brew sqlite installed, running `./configure --with-sqlite=no` I still see a single artefact of sqlite in config.log:
> 
> ```
> 5390:PKG_CONFIG_PATH='/usr/local/opt/qt/lib/pkgconfig:/usr/local/opt/sqlite/lib/pkgconfig:
> ```
> 
> It does not appear to impact the build, but I am curious how it's ending up there, as the code would suggest this would only be prepended to PKG_CONFIG_PATH when `if test "x$use_sqlite" != xno && $BREW list --versions sqlite3 >/dev/null` is satisfied?
> 
> That said, another tACK of [c932e0d](https://github.com/bitcoin/bitcoin/commit/c932e0d67e4b369e4265267da6c8bebac2b6fb53)

Hmm, did you run `./autogen.sh` before `./configure`?

willcl-ark commented:
> > With brew sqlite installed, running `./configure --with-sqlite=no` I still see a single artefact of sqlite in config.log:
> > ```
> > 5390:PKG_CONFIG_PATH='/usr/local/opt/qt/lib/pkgconfig:/usr/local/opt/sqlite/lib/pkgconfig:
> > ```
> > 
> > 
> > It does not appear to impact the build, but I am curious how it's ending up there, as the code would suggest this would only be prepended to PKG_CONFIG_PATH when `if test "x$use_sqlite" != xno && $BREW list --versions sqlite3 >/dev/null` is satisfied?
> > That said, another tACK of [c932e0d](https://github.com/bitcoin/bitcoin/commit/c932e0d67e4b369e4265267da6c8bebac2b6fb53)
> 
> Hmm, did you run `./autogen.sh` before `./configure`?

@hebasto you are right, re-running `autogen.sh` sorted it!

laanwj commented:
Code review ACK c932e0d67e4b369e4265267da6c8bebac2b6fb53

jonasschnelli reviewed (approved):
code review re-ACK c932e0d67e4b369e4265267da6c8bebac2b6fb53

MarcoFalke commented:
Backported in #20612 because this has been marked for backport
